src/datasets/arrow_dataset.py

```diff
@@ -425,17 +427,25 @@ def __setstate__(self, state):
         if self._indices is None and self._indices_data_files:
             self._indices = reader._read_files(self._indices_data_files)
 
-    def save_to_disk(self, dataset_path: str):
+    def save_to_disk(self, dataset_path: str, fs=None):
         """
-        Save the dataset in a dataset directory
+        Saves a dataset to a filesystem using either :class:`datasets.filesystem.S3FileSystem` or ``fsspec.spec.AbstractFileSystem``.
 
         Args:
-            dataset_path (``str``): path of the dataset directory where the dataset will be saved to
+            dataset_path (``str``): path (e.g. ``dataset/train``) or remote uri (e.g. ``s3://my-bucket/dataset/train``) of the dataset directory where the dataset will be saved to
+            fs (Optional[:class:`datasets.filesystem.S3FileSystem`,``fsspec.spec.AbstractFileSystem``],  `optional`, defaults ``None``): instance of :class:`datasets.filesystem.S3FileSystem` or ``fsspec.spec.AbstractFileSystem`` used to download the files from remote filesystem.
         """
         assert (
             not self.list_indexes()
         ), "please remove all the indexes using `dataset.drop_index` before saving a dataset"
         self = pickle.loads(pickle.dumps(self))
+
+        if is_remote_filesystem(fs):
+            dataset_path = preproc_dataset_path(dataset_path)
```

Comment:
In `save_to_disk`, when the dataset or the indices are in memory (as opposed to being loaded from arrow files on disk), then currently the arrow files are created in `dataset_path` using the ArrowWriter [here](https://github.com/huggingface/datasets/blob/85f02970f1872e9c394b76c37e7eaf2e0fb95bee/src/datasets/arrow_dataset.py#L453) for the dataset and [here](https://github.com/huggingface/datasets/blob/85f02970f1872e9c394b76c37e7eaf2e0fb95bee/src/datasets/arrow_dataset.py#L460) for the indices before being uploaded. However the ArrowWriter expects a local path. Currently if I understand correctly the arrow file is written in a local path, then copied to remote fs and then the local parent dir is deleted when you call 
```
        if "file" not in fs.protocol:
            shutil.rmtree(list(src.parents)[-2])
```

This can be improved on two points:
1. The local path that is used to create the ArrowWriter is currently the s3 path (see the 2 links to the code above). For example uploading a dataset that is in memory to `s3://foo` first creates an arrow file locally at `foo/xxxx.arrow`, then uploads it, and then the local `foo` is deleted. Maybe we can use a temp dir instead ?
2. Currently the local parent dir of the arrow file is always deleted (for remote fs). If the dataset is not in memory (i.e. is loaded from cached arrow files) then we don't want it to be deleted or the users might lose their data. First, this directory contains the arrow file of the current dataset and we don't want it to be deleted, and second this dir may also contain many other arrow files from other processing steps of this dataset or just other splits (train, test, val), as well as metadata files (json).

Reply:
1. So you would suggest creating a `temp_dir` as we have done it for `load_from_disk` [here](https://github.com/huggingface/datasets/blob/85f02970f1872e9c394b76c37e7eaf2e0fb95bee/src/datasets/arrow_dataset.py#L514)?

2. Yes it is only deleted if you save `files` to the remote filesystem. If I am correct this should be solved to if implement a `temp_dir` if the `filesystem` is "remote" ?

Reply:
1. I have looked into using `temp_dir` instead of the "local" version of the path. The problem is when we are serializing the state the "local" version of the path is used an `fs` determines if its uploaded to s3 or not. If want to work with a `temp_dir` we have to adjust this with conditional too. 
```python
with fs.open(Path(dataset_path).joinpath("state.json").as_posix(), "w", encoding="utf-8") as state_file:
    json.dump(state, state_file, indent=2, sort_keys=True)
with fs.open(
    Path(dataset_path).joinpath("dataset_info.json").as_posix(), "w", encoding="utf-8"
) as dataset_info_file:
    json.dump(dataset_info, dataset_info_file, indent=2, sort_keys=True)
``` 

2. I think what you explain is not a problem since we are saving the dataset to remote and before that, it is "saved" independently in a local representation `my-bucket/datasets/test` of the s3 uri `s3://my-bucket/datasets/test`. So i don´t think there is any risk in data loss

Reply:
About your two last points:
1. You can write the json files directly in the destination path using the filesystem. The local temp dir is only needed to write the dataset and the indices since the ArrowWriter works only for local paths.
2. Most of the time you don't need to write the dataset and the indices since they're already on disk. In this case it's the `data_files` attributes that tells you where the arrow files are located. By default the arrow files are in some cache directory of the lib along with many other arrow files. We have to make sure that we don't remove them because of `shutil.rmtree(list(src.parents)[-2])` so I would remove this line.

Reply:
1. but the problem is that `dataset_path` is used for both the ` Arrow_writer` and `json` files so need to add some sort of condition configuration for the `json`.  Sadly I cannot create a link to the code but it is in `arrow_dataset.py` line 482

Reply:
The temporary dir path can be used instead for the arrow_writer instead of the dataset_path. And dataset_path can be used for the json.

Reply:
yeah but not for `save_to_disk('local/path')` -> there is no `temp_dir`

Reply:
Indeed in the case of a local path this is the dest path that is used for the arrow_writer.

Another option would be to change that and always pass a temp dir path to the arrow_writer and then use `fs.put`, exactly like in the remote case.

Reply:
i have implemented `save_to_disk()`  using a `tmp_dir` no matter if it is a remote filesystem or a local one.